core/generate_index.go

```diff
@@ -261,291 +199,66 @@ func (ig *IndexGenerator) DropIndex(bucket []byte) error {
-}
-
-err := etl.Transform(
-	db,
-	dbutils.StorageChangeSetBucket,
-	dbutils.StorageHistoryBucket,
-	datadir,
-	startkey,
-	getExtractFunc(bytes2walker),
-	loadFunc)
-if err != nil {
-	return err
-}
-
-if err = SaveStageProgress(db, StorageHistoryIndex, blockNum); err != nil {
-	return err
-}
-
 
-
-
-
-
-
-
-
-
-bytes2walker := func(b []byte) changeset.Walker {
-		return changeset.AccountChangeSetBytes(b)
+	currentChunkKey := dbutils.IndexChunkKey(k, ^uint64(0))
+	indexBytes, err1 := state.Get(currentChunkKey)
+	if err1 != nil && err1 != ethdb.ErrKeyNotFound {
+		return fmt.Errorf("find chunk failed: %w", err1)
 	}
 
-	startkey := dbutils.EncodeTimestamp(blockNum)
+	currentIndex := dbutils.WrapHistoryIndex(indexBytes)
 
-	err := etl.Transform(
-		db,
-		dbutils.AccountChangeSetBucket,
-		dbutils.AccountsHistoryBucket,
-		datadir,
-		startkey,
-		getExtractFunc(bytes2walker),
-		loadFunc,
-	)
-	if err != nil {
-		return err
-	}
-*/
-
-func loadFunc(k []byte, valueDecoder etl.Decoder, state etl.State, next etl.LoadNextFunc) error { //nolint
-	var blockNumbers []uint64
-	err := valueDecoder.Decode(&blockNumbers)
-	if err != nil {
-		return err
-	}
-	for _, b := range blockNumbers {
-		vzero := (b & emptyValBit) != 0
-		blockNr := b &^ emptyValBit
-		currentChunkKey := dbutils.IndexChunkKey(k, ^uint64(0))
-		indexBytes, err1 := state.Get(currentChunkKey)
+	for i := 0; i < len(value); i += 9 {
+		b := binary.BigEndian.Uint64(value[i:])
+		vzero := value[i+8] == 1
+		blockNr := b
 		if err1 != nil && err1 != ethdb.ErrKeyNotFound {
 			return fmt.Errorf("find chunk failed: %w", err1)
 		}
-		var index dbutils.HistoryIndexBytes
-		if len(indexBytes) == 0 {
-			index = dbutils.NewHistoryIndex()
-		} else if dbutils.CheckNewIndexChunk(indexBytes, blockNr) {
+
+		if dbutils.CheckNewIndexChunk(currentIndex, blockNr) {
 			// Chunk overflow, need to write the "old" current chunk under its key derived from the last element
-			index = dbutils.WrapHistoryIndex(indexBytes)
-			indexKey, err3 := index.Key(k)
+			indexKey, err3 := currentIndex.Key(k)
 			if err3 != nil {
 				return err3
 			}
 			// Flush the old chunk
-			if err4 := next(indexKey, index); err4 != nil {
+			if err4 := next(indexKey, currentIndex); err4 != nil {
 				return err4
 			}
 			// Start a new chunk
-			index = dbutils.NewHistoryIndex()
-		} else {
-			index = dbutils.WrapHistoryIndex(indexBytes)
-		}
-		index = index.Append(blockNr, vzero)
-
-		err = next(currentChunkKey, index)
-		if err != nil {
-			return err
+			currentIndex = dbutils.NewHistoryIndex()
 		}
+		currentIndex = currentIndex.Append(blockNr, vzero)
 	}
+	err := next(currentChunkKey, currentIndex)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
 func getExtractFunc(bytes2walker func([]byte) changeset.Walker) etl.ExtractFunc { //nolint
 	return func(dbKey, dbValue []byte, next etl.ExtractNextFunc) error {
-		bufferMap := make(map[string][]uint64)
+		bufferMap := make(map[string][][]byte)
```

Comment:
I think for performance benefits we can pass `bufferMap` to `getExtractFunc` as an argument to avoid re-creating it on every `k/v` pair.